### ISSUE TITLE
RangeError: Maximum call stack size exceeded

### ISSUE THREAD

**Issue by benjamin-hg** [microsoft/TypeScript#9390]
While playing around with generic type constraints, I came across a bug in the typescriptServices.js used on the ts playground:

**TypeScript Version:**  1.8.2 (66d9c92bd3d6127ed3142876bc1a446bbd8395de)

**Code**

``` ts
function cast<TSparse extends TFull|TSparse, TFull extends TFull & TSparse>
(input:TSparse):TFull{
    return input; 
} 
```

**Expected behavior:**
A (meaningful) compile error or compiled code or both

**Actual behavior:** 
RangeError: "Maximum call stack size exceeded" and no compile result in TS Playground with this code.

**Comment by benjamin-hg** [comment]
I know it's an exotic kind of type constraint usage and I'm not very surprised it didn't work. I just reported it because the compiler probably is not supposed to crash.

**Comment by benjamin-hg** [comment]
BTW: what I was trying to do:

**Problem:**
I have an interface `Item` that has a _lot of non-optional properties_. In some unit tests, I need to create objects of that interface, but do not need all these properties.

I want to be able to define a "sparse" item with only some of the otherwise mandatory properties present.

**My approach:**
I created an interface `SparseItem` that has the _same properties, but all are optional_. I use it for the variable that holds my test dummy.

When I want to pass it to methods that expect an object of interface `Item`, I have to do a kind of cast. 
``` .ts
var item = <Item><any>sparseItem;
```
would have done the job, but not very safely. So I thought I could wirte a `cast` function that would do a better job. Usage like this:
``` .ts
var item:Item = cast<SparseItem, Item>(sparseItem);
```

**My goal:**
- When a property of `Item` is removed or renamed, but not on `SparseItem` and it is still used, I get an error.
- (Optional) When a property is added to `Item`, but not on `SparseItem`, I get an error.

**Comment by mhegazy** [comment]
This should be flagged as an error by the compiler. a type parameter can not depend on itself, this is not supported.
I think what you are looking for is something like a type operator that gives you an all optional version of a type. this is tracked by https://github.com/Microsoft/TypeScript/issues/7004

**Comment by benjamin-hg** [comment]
@mhegazy yes, exactly -- thanks for pointing me to #7004

**Comment by sandersn** [comment]
Fixed in 3.8
